api/environment.go

```diff
@@ -12,54 +11,93 @@ import (
+// an orchestrator defined in the Orchestration library or a direct local execution
+// of the r10k deploy environment command
 func (e EnvironmentController) DeployEnvironment(c *gin.Context) {
-	data := parsers.Data{}
-	h := helpers.Helper{}
-	cmd := exec.Command("r10k", "deploy", "environment")
-	conf := config.GetConfig().R10k
-	notify := config.GetConfig().ChatOps.Enabled
+	var data parsers.Data
+	var h helpers.Helper
+
+	// Set the base r10k command into a slice of strings
+	cmd := []string{"r10k", "deploy", "environment"}
+
+	// Get the configuration
+	conf := config.GetConfig()
+
+	// Setup chatops connection so we don't have to repeat the process
 	conn := chatopsSetup()
 
+	// Parse the data from the request and error if the parsing fails
 	err := data.ParseData(c)
 	if err != nil {
 		c.JSON(http.StatusInternalServerError, gin.H{"message": "Error Parsing Webhook", "error": err})
 		log.Errorf("error parsing webhook: %s", err)
 		c.Abort()
 		return
 	}
-	prefix := h.GetPrefix(data, conf.Prefix)
-	branch := h.GetBranch(data, conf.DefaultBranch)
-	env := h.GetEnvironment(branch, prefix, conf.AllowUppercase)
 
-	cmd.Args = append(cmd.Args, env)
-	cmd.Args = append(cmd.Args, fmt.Sprintf("--config=%s", h.GetR10kConfig()))
+	// Setup the environment for r10k from the configuration
+	env := h.GetEnvironment(conf.R10k.DefaultBranch, conf.R10k.Prefix, conf.R10k.AllowUppercase)
 
-	if conf.Verbose {
-		cmd.Args = append(cmd.Args, "-v")
+	// Append the environment and r10k configuration into the string slice `cmd`
+	cmd = append(cmd, env)
+	cmd = append(cmd, fmt.Sprintf("--config=%s", h.GetR10kConfig()))
+
+	// Set additional optional r10k options if they are set
+	if conf.R10k.Verbose {
+		cmd = append(cmd, "-v")
 	}
-	if conf.DeployModules {
-		cmd.Args = append(cmd.Args, "-m")
+	if conf.R10k.DeployModules {
+		cmd = append(cmd, "-m")
 	}
-	if conf.GenerateTypes {
-		cmd.Args = append(cmd.Args, "--generate-types")
+	if conf.R10k.GenerateTypes {
+		cmd = append(cmd, "--generate-types")
 	}
 
-	res, err := cmd.CombinedOutput()
-	if err != nil {
-		log.Errorf("cmd.Run() failed with error %s", string(res))
-		c.JSON(http.StatusInternalServerError, gin.H{"message": "error executing command", "error": string(res)})
-		c.Abort()
-		if notify {
-			conn.PostMessage(http.StatusInternalServerError, env)
+	// Determine if orchestration is enabled and either pass the cmd string slice to a
+	// the orchestrationExec function or localExec function.
+	// On an error this will:
+	//		* Log the error, orchestration type, and command
+	//		* Respond with an HTTP 500 error and return the command result in JSON format
+	//		* Abort the request
+	//		* Notify ChatOps service if enabled
+	//
+	// On success this will:
+	//		* Respond with an HTTP 202 and the result in JSON format
+	if conf.Orchestration.Enabled {
+		res, err := orchestrationExec(cmd)
+		if err != nil {
+			// TODO: Replace with proper custom error types
```

Comment:
Should this be done in this PR or later?

Reply:
After doing some research. This doesn't seem like an appropriate place for a custom error type. Primarily because those error types, more than likely, wouldn't be reused anywhere else.

Taking a page out of the `kubebuilder` book for API errors, using Errorf seems to be plenty appropriate here unless the code gets more complicated or we need to perform specific actions when an error is of a particular type.

Reply:
its fine for me :)